File: packages/core/android/src/main/java/com/datadog/reactnative/DdTrace.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.reactnative

import com.datadog.android.tracing.AndroidTracer
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import io.opentracing.Span
import io.opentracing.Tracer
import java.util.concurrent.TimeUnit


/**
 * The entry point to use Datadog's Trace feature.
 */
class DdTrace(reactContext: ReactApplicationContext, private val tracerProvider: () -> Tracer = { AndroidTracer.Builder().build() }) : ReactContextBaseJavaModule(reactContext) {

    private val spanMap: MutableMap<String, Span> = mutableMapOf()

    // lazy here is on purpose. The thing is that this class will be instantiated even
    // before Sdk.initialize is called, but Tracer can be created only after SDK is initialized.
    private val tracer by lazy { tracerProvider.invoke() }

    override fun getName(): String = "DdTrace"

    /**
     * Start a span, and returns a unique identifier for the span.
     * @param operation The operation name of the span.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the operation started (in milliseconds). If not provided, current timestamp will be used.
     */
    @ReactMethod
    fun startSpan(operation: String, context: ReadableMap, timestampMs: Long, promise: Promise) {
        val span = tracer.buildSpan(operation)
            .withStartTimestamp(TimeUnit.MILLISECONDS.toMicros(timestampMs.toLong()))
            .start()
        val spanContext = span.context()

        span.setTags(context.toHashMap())
        span.setTags(GlobalState.globalAttributes)
        val spanId = spanContext.toSpanId()
        spanMap[spanId] = span

        promise.resolve(spanId)
    }

    /**
     * Finish a started span.
     * @param spanId The unique identifier of the span.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the operation stopped (in milliseconds). If not provided, current timestamp will be used.
     */
    @ReactMethod
    fun finishSpan(spanId: String, context: ReadableMap, timestampMs: Long, promise: Promise) {
        val span = spanMap.remove(spanId)
        if (span == null) {
            promise.resolve(null)
            return
        }
        span.setTags(context.toHashMap())
        span.setTags(GlobalState.globalAttributes)
        span.finish(TimeUnit.MILLISECONDS.toMicros(timestampMs.toLong()))

        promise.resolve(null)
    }

    private fun Span.setTags(tags: Map<String, Any?>) {
        tags.forEach { (key, value) ->
            when (value) {
                is Boolean -> setTag(key, value)
                is Number -> setTag(key, value)
                is String -> setTag(key, value)
                else -> setTag(key, value?.toString())
            }
        }
    }
}

File: packages/core/ios/Sources/DdTrace.swift
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import Foundation
import Datadog

@objc(DdTrace)
class RNDdTrace: NSObject {

    @objc(requiresMainQueueSetup)
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    private lazy var tracer: OTTracer = tracerProvider()
    private let tracerProvider: () -> OTTracer
    private(set) var spanDictionary = [NSString: OTSpan]()

    internal init(_ tracerProvider: @escaping () -> OTTracer) {
        self.tracerProvider = tracerProvider
    }

    override convenience init() {
        self.init { Tracer.initialize(configuration: Tracer.Configuration()) }
    }
    @objc(methodQueue)
    let methodQueue: DispatchQueue = sharedQueue

    @objc(startSpan:withContext:withTimestampms:withResolver:withRejecter:)
    func startSpan(operation: String, context: NSDictionary, timestampMs: Int64, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        let id = UUID().uuidString as NSString
        let timeIntervalSince1970: TimeInterval = Double(timestampMs) / 1_000
        let startDate = Date(timeIntervalSince1970: timeIntervalSince1970)

        objc_sync_enter(self)
        spanDictionary[id] = tracer.startSpan(
            operationName: operation,
            childOf: nil,
            tags: castAttributesToSwift(context).mergeWithGlobalAttributes(),
            startTime: startDate
        )
        objc_sync_exit(self)

        resolve(id)
    }

    @objc(finishSpan:withContext:withTimestampms:withResolver:withRejecter:)
    func finishSpan(spanId: NSString, context: NSDictionary, timestampMs: Int64, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        objc_sync_enter(self)
        let optionalSpan = spanDictionary.removeValue(forKey: spanId)
        objc_sync_exit(self)

        if let span = optionalSpan {
            set(tags: castAttributesToSwift(context).mergeWithGlobalAttributes(), to: span)
            let timeIntervalSince1970: TimeInterval = Double(timestampMs) / 1_000
            span.finish(at: Date(timeIntervalSince1970: timeIntervalSince1970))
        }
        
        resolve(nil)
    }

    private func set(tags: [String: Encodable], to span: OTSpan) {
        for (key, value) in tags {
            span.setTag(key: key, value: value)
        }
    }
}
